optaplanner-wb-screens/optaplanner-wb-solver-editor/optaplanner-wb-solver-editor-client/src/main/java/org/optaplanner/workbench/screens/solver/client/editor/TerminationTreeItemContent.java

```diff
@@ -15,84 +15,408 @@
+    public Widget asWidget() {
+        return view.asWidget();
+    }
+
+    public TerminationTreeItemContentView getView() {
+        return view;
+    }
 
-    public TerminationTreeItemContent( TerminationConfigForm terminationConfigForm, TreeItem treeItem, TerminationConfigOption terminationConfigOption, TerminationConfigModel model, String value ) {
+    public void setModel( TerminationConfigModel model ) {
         this.model = model;
+    }
+
+    public void setTerminationConfigOption( TerminationConfigOption terminationConfigOption ) {
+        this.terminationConfigOption = terminationConfigOption;
+        view.setNestedTreeItem( terminationConfigOption == NESTED );
+        if ( terminationConfigOption == NESTED ) {
+            view.setDropDownHelpContent( I18N_CONSTANTS.TerminationCompositionStyleHelp() );
+        }
+        TerminationManager operation = terminationOperationMap.get( terminationConfigOption );
+        if ( operation != null ) {
+            operation.setLabelStrings();
+        }
+        for ( TerminationConfigOption option : TerminationConfigOption.values() ) {
+            if ( terminationConfigOption == option ) {
+                continue;
+            }
+            TerminationManager terminationManager = terminationOperationMap.get( option );
+            if ( terminationManager != null ) {
+                terminationManager.hideViewInputs();
+            }
+        }
+    }
+
+    public void setTreeItem( TreeItem treeItem ) {
         this.treeItem = treeItem;
+        view.setRoot( treeItem.getParentItem() == null );
+    }
 
-        boolean root = treeItem.getParentItem() == null;
+    public void onTerminationTypeSelected( String terminationType ) {
+        terminationConfigForm.addNewTerminationType( terminationType, this );
+    }
 
-        terminationTypeLabel.setText( terminationConfigForm.getModelManager( terminationConfigOption ).getLocalizedTerminationConfigOption() );
-        terminationTypeLabel.setVisible( !TerminationConfigOption.NESTED.equals( terminationConfigOption ) );
+    public void setTerminationConfigForm( TerminationConfigForm terminationConfigForm ) {
+        this.terminationConfigForm = terminationConfigForm;
+    }
 
-        terminationValueLabel.setText( value );
-        terminationValueLabel.setVisible( !root && !TerminationConfigOption.NESTED.equals( terminationConfigOption ) );
+    public void removeTreeItem() {
+        TerminationTreeItemContent parent = ( treeItem.getParentItem() == null ? this : (TerminationTreeItemContent) treeItem.getParentItem().getUserObject() );
+        if ( terminationConfigOption == NESTED ) {
+            parent.getModel().getTerminationConfigList().remove( model );
+            if ( parent.getModel().getTerminationConfigList().isEmpty() ) {
+                parent.getModel().setTerminationConfigList( null );
+            }
+        } else {
+            terminationOperationMap.get( terminationConfigOption ).removeModelValue();
```

Comment:
@ge0ffrey IMO NPE's at runtime in compiled JavaScript code do not stick out like a sore thumb - unless you choose to monitor the browser's developer console and even then the stack trace is highly obfuscated... that said we don't tend to program overly defensively for the client on the whole anyway.